ambry-network/src/main/java/com.github.ambry.network/BlockingChannelConnectionPool.java

```diff
@@ -223,6 +223,15 @@ public void destroyBlockingChannel(BlockingChannel blockingChannel) {
       synchronized (lock) {
         // decrement the number of connections to the host and port. we were not able to maintain the count
         numberOfConnections.decrementAndGet();
+        // at this point we are good to clean up the available connections since re-creation failed
+        do {
```

Comment:
I don't think taking a write lock on the `rwLock` specifically helps with anything. With the `poll()` and `remove()` change, you ensure that anything that is available at that point in time gets destroyed. Sure a `releaseBlockingChannel()` could come in later, but that can happen even if the write lock was taken.

I think the scenario is uncommon - it can only happen for those connections that are checked in after an earlier connection fails. And in the case, the original issue that we are solving might turn up - which may be fine since the scenario is uncommon. If not, this will have to be dealt with in a different way.